warehouse/keyword-common/src/main/java/datawave/util/keyword/YakeKeywordExtractor.java

```diff
@@ -0,0 +1,718 @@
+     * @param maxNGrams the maximum ngram size.
+     * @return a list of ngrams generated from the input sentences ordered based on where they appeared in the input
+     *    sentences. These ngrams are not deduplicated and the words in these ngrams are not normalized in any way.
+     */
+    protected static List<List<String>> calculateNGrams(List<List<String>> sentences, int minNGrams, int maxNGrams) {
+        return sentences.stream().flatMap(tokens -> // for each sentence
+        IntStream.range(0, tokens.size()) // for each token
+                .mapToObj(start -> IntStream.range(minNGrams, maxNGrams + 1) // for each window size
+                        .mapToObj(size -> {
+                            if (size > tokens.size()) {
+                                // if the window size is larger than the number of tokens in the sentence, we'll at
+                                // least return a range covering the existing tokens, this mimics scala's slide()
+                                // behavior.
+                                return IntStream.range(start, tokens.size());
+                            }
+
+                            // otherwise, if there are enough tokens remaining to generate a full window,
+                            // return a range covering the next window.
+                            final int end = start + size;
+                            if (end <= tokens.size()) {
+                                return IntStream.range(start, end);
+                            }
+
+                            // otherwise there are not enough tokens to form a full window, return nothing.
+                            return IntStream.empty();
+                        })) // return a stream of streams of all windows
+                .reduce(Stream.empty(), Stream::concat)// reduce them to a single stream
+                .map(k -> k.mapToObj(tokens::get) // get a string containing the tokens for each position in each list of positions.
+                        .collect(Collectors.toList()))
+                .filter(s -> !s.isEmpty())).collect(Collectors.toList());
+    }
+
+    /** Build a co-occurrence matrix of words based on a set of ngrams. Words are down-cased prior to adding them
+     *  to the matrix. We track the max sentence of co-occurrence instances for each pair of words.
+     *
+     * @param ngrams the set of ngrams to analyze
+     * @param left if true, the matrix is based on preceding words (e.g., those that appear to the left of the target word.)
+     *             if false, the matrix is based on following words (e.g., those that appear to the right of the target word.)
+     * @return a map where the key is a word and the value is the list of words that co-occur with that word. For each
+     *   co-occurrence, we track the max sentence id where that co-occurrence happened.
+     */
+    protected Map<String, Map<String, Integer>> buildCoOccurrence(List<List<String>> ngrams, boolean left) {
```

Comment:
ngrams I don't believe have been normalized at this point. They probably need to be?

Reply:
we spoke about puncuation specifically